cellengine/resources/gate.py

```diff
@@ -1,158 +1,74 @@
-                be made to find the file by name. If zero or more than one file exists
-                with the name, an error will be thrown. Looking up files by name is
-                slower than using the ID, as this requires additional requests
-                to the server. If specified, do not specify ``fcs_file_id``.
-            locked (bool): Prevents modification of the gate via the web interface.
-            create_population (bool): Automatically create corresponding population.
-        """
-        if type(gates) is list:
-            return [cls._build_gate(gate) for gate in gates]
+    def update(self):
+        """Save changes to this Gate to CellEngine.
+        If this gate does not exist, it will be created."""
+        if not self._id or not is_valid_id(self._id):
+            gate = self.client.create(self)
         else:
-            return cls._build_gate(gates)
-
-    @classmethod
-    def _build_gate(cls, gate):
-        """Get the gate type and return instance of the correct subclass."""
-        module = importlib.import_module(__name__)
-        gate, _ = cls._separate_gate_and_population(gate)
-        gate_type = getattr(module, gate["type"])
-        return gate_type(properties=gate)
+            gate = self.client.update(self)
+        self.__setstate__(gate.__getstate__())  # type: ignore
 
     @classmethod
-    def _separate_gate_and_population(cls, gate):
-        try:
-            if "gate" in gate.keys():
-                return gate["gate"], [k for k in gate.keys() if k != "gate"]
-            else:
-                return gate, None
-        except KeyError:
-            raise ValueError("Gate payload format is invalid")
-
-    @staticmethod
-    def delete_gates(
-        experiment_id, _id: str = None, gid: str = None, exclude: str = None
-    ) -> None:
-        """Deletes a gate or a tailored gate family.
-
-        Specify the top-level gid when working with compound gates (specifying
-        the gid of a sector (i.e. one listed in ``model.gids``) will result in
-        no gates being deleted). If ``_id`` is specified, only that gate will be
-        deleted, regardless of the other parameters specified. May be called as
-        a static method on `cellengine.Gate` or on an `Experiment` instance.
-
-        Args:
-            experiment_id (str): ID of experiment.
-            _id (str): ID of gate.
-            gid (str): ID of gate family.
-            exclude (str): Gate ID to exclude from deletion.
-
-        Example:
-            ```python
-            cellengine.Gate.delete_gates(experiment_id, gid = <gate family ID>)
-            # or
-            experiment.delete_gates(_id = <gate ID>)
-            # or
-            experiment.delete_gates(gid = <gate family ID>, exclude = <gate ID>)
-            ```
+    def factory(cls, gate):
+        """Create a gate from a dict of keys."""
+        if "_id" not in gate.keys():
+            gate["_id"] = None
```

Comment:
This method also only exists for Gates I think.

When do we need this `_id = None` default?

Reply:
`cattrs` still needs a default value for `_id`, since it's a required property of a `Gate`. Alternatively, I could make it an optional property, which looks like:

```python
class Gate:
    experiment_id: str = field(on_setattr=readonly)
    type: str
    x_channel: str
    model: Dict
    _id: Optional[str] = field(on_setattr=readonly, default=None)
    # ...

    @classmethod
    def factory(cls, gate):
        """Create a gate from a dict of keys."""
        return converter.structure(gate, Gate)
    # ...
```

Reply:
I'm still confused why this class needs it but, say, Compensation doesn't. It looks like `factory()` is only used by tests. Would normal users need it? If no, then how about setting the ID in the test code and just using `from_dict()`?

Reply:
This is a bit of a holdover from how I had previously set up multiple-gate creation, and I agree that it can be deleted.

If I use `from_dict`, I either need to...

Alter `from_dict` to be like the deleted `factory` method:
```python
class Gate:
    ...
    def from_dict(cls, data: dict):
        if "_id" not in data.keys():
            data["_id"] = None
        return converter.structure(data, cls)
```

Or create gates in tests like `RectangleGate.create(*args)`. This is how it's supposed to be used, so I guess this is fine. The only feature we lose is gate creation from a `dict`, which requires the `if "_id"` syntax above.

Reply:
`Gate` needs this logic because it's the only class where an object can exist that does not correspond to an entity in CellEngine. An `_id` must come from CE (with the exception of creating new i.e. quadrant gates, which need an id for each of the component gates)

Reply:
> `Gate` needs this logic because it's the only class where an object can exist that does not correspond to an entity in CellEngine.

Ah okay, this is the important part, and in another comment you say that's because the bulk API uses `client.create([gate1, gate2])`.

Is the only way for the user to use the bulk API to do `gate1 = RectangleGate()`, `gate2 = ...`, `client.create([gate1, gate2])`? If we instead had a `Gate.create_many(list_of_dicts)`, would that resolve both this inconsistency and https://github.com/primitybio/cellengine-python-toolkit/pull/136/files#r774806702 <- that one?

Reply:
That _is_ the only way to create multiple gates, currently. I previously had a method for creating many gates from a list of dicts, but it seems much easier to do `gate1 = RectangleGate.create(), gate2 = ...` than to create a correctly-formed dict (i.e. `gate1 = {experiment_id: ..., model = { ... } }`). I think the options are:

- force the user to use the `<Type>Gate.create()` methods, then `client.create([g1, g2, g3])` (keep things as they are)
- have a `Gate.create_many([dict or Gate])` and possibly...
- ...expose the logic inside <Type>Gate.create as `Gate.format_<type>_gate() # returns dict` and use that in the proposed `Gate.create_many()`, or....
- ...split gate formatting and creation inside each class, like this:
```python
class RectangleGate(Gate):
    @classmethod
    def create(cls, *args, **kwargs):
        return ce.APIClient().create(cls.format(*args, **kwargs))

    @classmethod
    def format( ... )
        return Gate ( ... )

```

Reply:
If we did one of the latter two, then `<Type>Gate.create` would always return a gate entity, `.format` would always return a `dict`, and the user would have to call `APIClient.create()` (or `Gate.create`?) to post the gate. 

This would eliminate the `if not self._id or not is_valid_id(self._id):` logic, which I think handles both inconsistencies

Reply:
I think there's another option that's easier than those: `Gate.create_many([dict1, dict2])` would automatically map the create/format method over those raw/partial dicts as if `Gate.create()` had been called on each of them individually.

Ideally we'd have both `Gate.create_many()`, which would use the `type` property of each dict to determine the type, and `<Type>.create_many()`, which would add `type: "<Type>Gate"`.